### ISSUE TITLE
Clarity For Stable Diffusion Training / Inference

### ISSUE THREAD

**Issue by ExponentialML** [rinongal/textual_inversion#7]
Hello! Thanks for your work. I was reading through your paper, and had some questions.

Initially when I tried this, I was having trouble with the default parameters listed [here](https://github.com/rinongal/textual_inversion#inversion). It turned out I was getting a sculpture representation of the images images I was trying to invert! 🙃

I then realized, that in the `v1-finetune.yaml` has a string list of `initializer_words`, and not `init_word`.
My understanding (correct me if I'm wrong) is that the words are converted into a numerical representation of the images you are trying to invert, and then the *pseudo word* is the representation of the inversion.

In the paper, it says that one token is sufficient with roughly 5K iterations. My confusion stems from `placeholder_strings` and `initializer_words`. 

The `placeholder_strings` is a pseudo keyword to that describes the subject, and can be anything. Am I right to assume that the `initializer_words` *guide* the image you're trying to invert? Is there any limit to this during training? Would over describing lead to some sort of overfitting edge case?

I know that you're still working on support for SD, but I'm eager to try during development. Any clarity would be greatly appreciated!

**Comment by rinongal** [comment]
The placeholder string (default "*") is the word whose token will be replaced by the new, learned embedding in the text. The initializer words are used to initialize the new embedding for this string, which is then further optimized during training.

As the paper notes, we had the best results when using only a single placeholder string (i.e. learning just a single embedding). The code itself supports changing to multiple placeholders (if you want to learn several concepts together) or multiple embedding vectors per concept, but these are off by default.

The initializer words don't actually act as extra placeholders, they just tell the model how to initialize the new embeddings. The only limit is that they need to be a single token string (that is, they need to match a single embedding vector). Unfortunately what counts as a single token depends on the model so I can't give you tips there, but the model will exit with an appropriate error if your string is more than a single token.

**Comment by rinongal** [comment]
Also - if you provide an init word as an argument to the training script, it will simply overwrite the value from the config.

**Comment by ExponentialML** [comment]
Awesome, thanks for clearing that up. I understand it clearly and I'm also able to verify. Cheers!

**Comment by 1blackbar** [comment]
> Awesome, thanks for clearing that up. I understand it clearly and I'm also able to verify. Cheers!

do You have white conditioning gs images as well ? also does it restart to new folder every 2500 iterations ?

**Comment by ExponentialML** [comment]
Hello again. Instead of starting a new issue, I'll re-open this one and close once resolved.

I've been messing around with different parameters and options, and had further questions that I'm curious about.

1 - Increasing the `num_vectors_per_token` to higher amounts seems to create better correlation between the training images and sampled ones. This seems to offset the `scale` parameter during inference though, leading extreme bias (overfitting perhaps?) towards the `.pt` file.

In many cases, this allows for retaining features and structures of the training set. Is there a reason as to why that is, and is there a way to alleviate the strength of the `.pt` file to find a balance between prompt and the inversion? Would it be better to use a different approach such as using `per_image_tokens` along with `progressive_words`?

2 - I can't find any reference as to what the `embedding_reg_weight` parameter does. Could you possibly explain what this parameter affects?

3 - Does the `imagenet_templates_small` prompt list actually matter during training, or is it only used for validation?

Thanks again for your help!

---

@1blackbar, The white conditioning images are normal (just displays the prompt)  and I've never had an issue with restarting a new folder every iteration. If there are any further questions on the topic, it may be a good idea to start a separate issue.

**Comment by XodrocSO** [comment]
@ExponentialML Good questions! I'm curious about the imagenet prompt list too!  Also noticed the one that's apparently for 2 subjects? `imagenet_dual_templates_small`  'a photo of a {} with {}',

**Comment by rinongal** [comment]
As you noted, extra vectors per token let you better reconstruct the subject, but they harm editability. Basically you're predicting things that are further and further out of distribution for the model so it has a hard time reasoning over them (+ more words typically means more weight in the prompt). If you want to play with the weight of the .pt prompt, the best way we found is to just change the learning rate during training. Higher LR leads to better match for the original set, at the cost of some editability. Lower LR will do the opposite. You can find some discussion on this in the paper.

embedding_reg_weight is a regularization loss that tries to keep the learned embeddings close to the word you used to initialize them, kind of a way to keep things close to the real world distribution. But if you turn it on, you'll see that the method fails to capture the concepts in a satisfying manner.

Template list is used during training, yes. It helps generalization to new prompts.

The one for multiple subjects is used when we use the per-image tokens. Basically one of those placeholders will be filled by the global placeholder ("*") while the other will be filled by an image-specific placeholder.

You can find more details on all of these things in the paper.

**Comment by ExponentialML** [comment]
You're the man. Thank you so much!
